### PR TITLE
Modification in transport pk to support seawater intrusion

### PR DESCRIPTION
**Pull request by dasvyat**
The major modification is setting total component concentration as the primary evaluator.
The density model for salt water is created.

### CODE REVIEW

File: src/pks/flow/constitutive_relations/wrm/wrm_van_genuchten.cc [hunk 1]
```diff
@@ -205,5 +205,47 @@ void WRMVanGenuchten::InitializeFromPlist_() {
...
+/* ******************************************************************
+* Suction formula: input is liquid saturation.
+****************************************************************** */
+double WRMVanGenuchten::suction_head(double s) {
+ 
+  double se = (s - sr_)/(1-sr_);
+  //if (function_ == FLOW_WRM_MUALEM) {
+  if (se > FLOW_WRM_TOLERANCE) {
+    return -(1./alpha_) * pow(pow(se, -1./m_) - 1, 1. - m_);
+  }else{
```
Comment (ecoon) [inline src/pks/flow/constitutive_relations/wrm/wrm_van_genuchten.cc:217]: ```} else {```

Please remove comments and blank lines to match the formatting standards.

File: src/pks/flow/overland_pressure_pk.cc [hunk 2]
```diff
@@ -1191,7 +1191,9 @@ void OverlandPressureFlow::FixBCsForOperator_(const Teuchos::Ptr<State>& S,
             " Aff "<< Aff[f](0,0) << "\n";
         }
       }
+
```
Comment (ecoon) [inline src/pks/flow/overland_pressure_pk.cc:1194]: Again, I'd suggest using emacs whitespace-cleanup or comparable to remove spaces in empty lines, and there is no need to add blank lines all over the place.  The formatting norms for both Amanzi and ATS are 1 blank line to break up logical breaks in code, and 2 blank lines between function definitions.

File: src/pks/flow/constitutive_relations/wrm/wrm_van_genuchten.cc [hunk 3]
```diff
@@ -205,5 +205,47 @@ void WRMVanGenuchten::InitializeFromPlist_() {
...
+    return -1e+10;
+  }
+
+ 
+}
+
+
+/* ******************************************************************
+ * D Relative permeability / D capillary pressure pc.
```
Comment (ecoon) [inline src/pks/flow/constitutive_relations/wrm/wrm_van_genuchten.cc:226]: fix the documentation.  Can you also add a comment on what "suction head" is?  I've never heard of it!

File: src/pks/flow/richards_pk.cc [hunk 4]
```diff
@@ -418,13 +420,28 @@ void Richards::SetupPhysicalEvaluators_(const Teuchos::Ptr<State>& S) {
...
+      Teuchos::ParameterList wrm_plist(plist_->sublist("water retention evaluator"));
+      wrm_plist.setName(suction_key_);
+      wrm_plist.set("field evaluator type", "WRM suction head");
+      S->FEList().set(suction_key_, wrm_plist);
+
+      //-- suction
+      S->RequireField(suction_key_)->SetMesh(mesh_)->SetGhosted()
+        ->AddComponent("cell", AmanziMesh::CELL, 1);
+      S->RequireFieldEvaluator(suction_key_);
+    }
```
Comment (ecoon) [inline src/pks/flow/richards_pk.cc:435]: Wait wait, I don't like this.  What is suction, and why is every Richards PK going to have to add a new vector and new evaluator for it?  Nowhere else in richards_* is this key used, so I don't understand what you are doing with it.  If you need it for another PK, please add it there, not in Richards PK.

File: src/pks/mpc/mpc_flowreactivetransport_pk.cc [hunk 5]
```diff
@@ -153,10 +153,11 @@ bool FlowReactiveTransport_PK_ATS::AdvanceStep(double t_old, double t_new, bool
 
   if (std::abs(t_old + dt_done - t_new) / (t_new - t_old) < 0.1*min_dt_) {
     // done, success
-    if (vo_->getVerbLevel() >= Teuchos::VERB_EXTREME) *vo_->os()<<"Slave step is successful\n";
+    if (vo_->getVerbLevel() >= Teuchos::VERB_HIGH)
+      *vo_->os()<<"Slave step is successful (use "<<ncycles<<" subcycles.)"<<"\n";
     return false;
   } else {
-    if (vo_->getVerbLevel() >= Teuchos::VERB_EXTREME) *vo_->os()<<"Slave step is failed\n";
+    if (vo_->getVerbLevel() >= Teuchos::VERB_HIGH) *vo_->os()<<"Slave step is failed\n";
```
Comment (ecoon) [inline src/pks/mpc/mpc_flowreactivetransport_pk.cc:160]: >= is not guaranteed by Trilinos, and the fact that these are ordered and are integers is not guaranteed.  Please prefer to use the VerboseObject interface: `vo_->os_OK(Teuchos::VERB_EXTREME)`

File: src/pks/transport/transport_ats.hh [hunk 6]
```diff
@@ -94,6 +94,7 @@ public:
 
   virtual std::string name() { return "transport_ats"; }
 
+
```
Comment (ecoon) [inline src/pks/transport/transport_ats.hh:97]: more unnecessary whitespace

File: src/pks/transport/transport_ats_pk.cc [hunk 7]
```diff
@@ -52,6 +53,17 @@ Transport_ATS::Transport_ATS(Teuchos::ParameterList& pk_tree,
   PK(pk_tree, glist, S, solution),
   PK_PhysicalExplicit<Epetra_Vector>(pk_tree, glist, S, solution)
 {
+
+
+  key_ = Keys::readKey(*plist_, domain_, "primary variable", "total_component_concentration");
+  tcc_key_ = key_;
+  
+  // set up the primary variable solution, and its evaluator
+  Teuchos::ParameterList& FElist = S->FEList();
```
Comment (ecoon) [inline src/pks/transport/transport_ats_pk.cc:62]: Oh wow, was TCC not already a primary variable evaluator?  I guess not.  Glad this is happening!

Comment (ecoon) [inline src/pks/transport/transport_ats_pk.cc:62]: Instead of doing this way, please use:

```
Teuchos::ParameterList& pv_sublist = S->GetEvaluatorList(key_);
```
https://github.com/amanzi/amanzi/blob/master/src/state/State.hh#L299

This is more robust because it allows for domain set names, which we use extensively in transport for subgrid models.

File: src/pks/transport/transport_ats_pk.cc [hunk 8]
```diff
@@ -52,6 +53,17 @@ Transport_ATS::Transport_ATS(Teuchos::ParameterList& pk_tree,
...
+  Teuchos::ParameterList& pv_sublist = FElist.sublist(key_);
+  pv_sublist.set("evaluator name", key_);
```
Comment (ecoon) [inline src/pks/transport/transport_ats_pk.cc:64]: "evaluator name" is no longer used for anything.  So you don't need to set this.  The name is assumed to be the ParameterList::name() value.

File: src/pks/transport/transport_ats_pk.cc [hunk 9]
```diff
@@ -129,6 +142,7 @@ void Transport_ATS::Setup(const Teuchos::Ptr<State>& S)
   dissolution_ = plist_->get<bool>("allow dissolution", false);
   max_tcc_ = plist_->get<double>("maximum concentration", 0.9);
 
+
```
Comment (ecoon) [inline src/pks/transport/transport_ats_pk.cc:145]: haha, so much white space :D

File: src/pks/transport/transport_ats_pk.cc [hunk 10]
```diff
@@ -1348,6 +1375,9 @@ void Transport_ATS::AdvanceDonorUpwind(double dt_cycle)
         tcc_next[i][c] = 0.;
       }
     }
+    // if ((domain_name_=="surface")&&(rank==1)&&(c > 47)&&(c<53)){
+    //   std::cout<<c<<" : "<< vol_phi_ws_den<<" "<<(*ws_end)[0][c]<<" "<< (*mol_dens_end)[0][c]<<" "<<(*conserve_qty_)[0][c]<<" tcc "<<tcc_next[0][c]<<"\n";
+    // }
```
Comment (ecoon) [inline src/pks/transport/transport_ats_pk.cc:1380]: remove please

**Review by ecoon — CHANGES_REQUESTED** [review]
Lots of whitespace and formatting to clean up.

More importantly, I don't like the addition of a new variable to Richards equation that doesn't appear to be used anywhere.  Can you explain what it is used for?  And it almost certainly shouldn't be required in the Richards PK if it isn't used in the Richards PK.

File: src/pks/transport/transport_ats_pk.cc (hunk 10 above)
Comment (dasvyat) [reply src/pks/transport/transport_ats_pk.cc:1380]: Removed

File: src/executables/CMakeLists.txt [hunk 11]
```diff
@@ -48,9 +48,9 @@ include_directories(${AMANZI_BINARY_DIR}) # required to pick up amanzi_version.h
 include_directories(${ATS_BINARY_DIR})
 
 if (ENABLE_FATES)
-  link_directories(/nh/u/dasvyat/Coastal/ats-fates-new/install/lib/)
-  link_directories(/nh/u/dasvyat/Coastal/ParallelIO/install/lib/)
-  link_directories(/nh/u/dasvyat/Coastal/pnetcdf-1.11.2/install/lib)
+  # link_directories(/nh/u/dasvyat/Coastal/ats-fates-new/install/lib/)
+  # link_directories(/nh/u/dasvyat/Coastal/ParallelIO/install/lib/)
+  link_directories(${CMAKE_PREFIX_PATH}/lib/)
```
[the remark this answers is not in the document]
Comment (dasvyat) [reply src/executables/CMakeLists.txt:53]: Removed FATES linking

File: src/executables/CMakeLists.txt [hunk 12]
```diff
@@ -151,7 +151,8 @@ if (APPLE AND BUILD_SHARED_LIBS)
 endif()			  
 
 if (ENABLE_FATES)
-  set(fates_link_libs -lfates -lparteh -lshared  -lpiof -lpioc -lpnetcdf -lgptl)
+  set(fates_link_libs -lfates -lparteh -lshared  -lpiof -lpioc -lpnetcdf)
+#  set(fates_link_libs -lfates -lparteh -lshared  -lpiof -lpioc -lpnetcdf -lgptl)
```
[the remark this answers is not in the document]
Comment (dasvyat) [reply src/executables/CMakeLists.txt:155]: Removed conditional FATES linking

File: src/pks/biogeochemistry/CMakeLists.txt [hunk 13]
```diff
@@ -20,6 +19,12 @@ set(ats_bgc_src_files
   constitutive_models/carbon/bioturbation_evaluator.cc  
   )
 
+if (ENABLE_FATES)
+  list (APPEND ats_bgc_src_files fates/fates_pk.cc)
+endif()
+
+
+
```
[the remark this answers is not in the document]
Comment (dasvyat) [reply src/pks/biogeochemistry/CMakeLists.txt:27]: Removed

File: src/pks/flow/constitutive_relations/wrm/wrm_van_genuchten.cc [hunk 14]
```diff
@@ -205,5 +205,47 @@ void WRMVanGenuchten::InitializeFromPlist_() {
...
+    dpsidse = ((m_ - 1)/(alpha_ * m_ * se)) * pow(x - 1, -m_) * x;
+  }else{
+    dpsidse = 0.;
+  }
+  return -dpsidse / (1 - sr_);
+
+
+}
+  
+
```
[the remark this answers is not in the document]
Comment (dasvyat) [reply src/pks/flow/constitutive_relations/wrm/wrm_van_genuchten.cc:248]: Fully agreed

File: src/pks/biogeochemistry/bgc_simple/bgc_simple.cc [hunk 15]
```diff
@@ -517,6 +517,16 @@ void BGCSimple::FieldToColumn_(AmanziMesh::Entity_ID col, const Epetra_Vector& v
   }
 }
 
+// helper function for pushing field to column
+void BGCSimple::FieldToColumn_(AmanziMesh::Entity_ID col, const Epetra_Vector& vec,
+                               double* col_vec, int ncol) {
+  ColIterator col_iter(*mesh_, mesh_surf_->entity_get_parent(AmanziMesh::CELL, col), ncol);
+  for (std::size_t i=0; i!=col_iter.size(); ++i) {
+    col_vec[i] = vec[col_iter[i]];
```
Comment (ecoon) [inline src/pks/biogeochemistry/bgc_simple/bgc_simple.cc:525]: this could just be:

    auto& col_iter = mesh_->cells_of_column(col);
    for (...) {
      col_vec[i] = vec[col_iter[i]]
    }

And remove the ColIterator object which is not needed.

File: src/pks/flow/richards_pk.cc [hunk 16]
```diff
@@ -86,6 +86,7 @@ Richards::Richards(Teuchos::ParameterList& pk_tree,
   sat_gas_key_ = Keys::readKey(*plist_, domain_, "saturation gas", "saturation_gas");
   sat_ice_key_ = Keys::readKey(*plist_, domain_, "saturation ice", "saturation_ice");
 
+
```
Comment (ecoon) [inline src/pks/flow/richards_pk.cc:89]: Can you just revert richards_pk.cc to what it is on master, e.g.

    git checkout origin/master -- src/pks/flow/richards_pk.cc

No need for any changes to this PK.

File: src/pks/flow/overland_pressure_ti.cc [hunk 17]
```diff
@@ -327,5 +326,6 @@ void OverlandPressureFlow::UpdatePreconditioner(double t,
   iter_++;
 };
 
+
 }  // namespace Flow
 }  // namespace Amanzi
```
Comment (ecoon) [inline src/pks/flow/overland_pressure_ti.cc:331]: See comment below on Richards.  No need for changes to these files, and they just make later pull requests/merges with other branches harder.

File: src/pks/mpc/mpc_coupled_transport.cc [hunk 18]
```diff
@@ -97,6 +97,7 @@ bool CoupledTransport_PK::AdvanceStep(double t_old, double t_new, bool reinit)
   bool fail = false;
```
Comment (ecoon) [inline src/pks/mpc/mpc_coupled_transport.cc:97]: see above, no changes needed.

File: src/pks/transport/sediment_transport/sediment_transport_pk.cc [hunk 19]
```diff
@@ -119,6 +119,7 @@ void SedimentTransport_PK::Setup(const Teuchos::Ptr<State>& S)
 
   water_tolerance_ = tp_list_->get<double>("water tolerance", 1e-1);
   max_tcc_ = tp_list_->get<double>("maximal concentration", 0.9);
+  sediment_density_ = tp_list_->get<double>("sediment density [kg/m^3]");
```
Comment (ecoon) [inline src/pks/transport/sediment_transport/sediment_transport_pk.cc:122]: Can you make this "sediment density [kg m^-3]".  Some of our xml parsing chokes on "/" in names, so all ATS unit strings at this point have been changed to negative exponents rather than divide (which is a bit confusing from an order of operations standpoint anyway, as it requires parentheses in some cases).

File: src/pks/transport/sediment_transport/settlement_evaluator.cc [hunk 20]
```diff
@@ -26,7 +26,9 @@ SettlementRateEvaluator :: SettlementRateEvaluator(Teuchos::ParameterList& plist
   tau_d_ = plist_.get<double>("critical shear stress");
   ws_ = plist_.get<double>("settling velocity");
   gamma_ = plist_.get<double>("specific weight of water");
+  sediment_density_ = plist_.get<double>("sediment density [kg/m^3]");
```
Comment (ecoon) [inline src/pks/transport/sediment_transport/settlement_evaluator.cc:29]: See above, [kg m^-3]

File: src/pks/transport/sediment_transport/trapping_evaluator.cc [hunk 21]
```diff
@@ -32,6 +32,7 @@ TrappingRateEvaluator :: TrappingRateEvaluator(Teuchos::ParameterList& plist) :
   alpha_ = plist_.get<double>("alpha");
   beta_ = plist_.get<double>("beta");
   gamma_ = plist_.get<double>("gamma");
+  sediment_density_ = plist_.get<double>("sediment density [kg/m^3]");
```
Comment (ecoon) [inline src/pks/transport/sediment_transport/trapping_evaluator.cc:35]: See above

File: src/pks/transport/transport_ats_pk.cc [hunk 22]
```diff
@@ -52,6 +53,15 @@ Transport_ATS::Transport_ATS(Teuchos::ParameterList& pk_tree,
   PK(pk_tree, glist, S, solution),
   PK_PhysicalExplicit<Epetra_Vector>(pk_tree, glist, S, solution)
 {
+
+
```
Comment (ecoon) [inline src/pks/transport/transport_ats_pk.cc:57]: extra spaces

File: src/pks/transport/transport_ats_pk.cc [hunk 23]
```diff
@@ -111,6 +121,7 @@ void Transport_ATS::set_states(const Teuchos::RCP<State>& S,
 ****************************************************************** */
 void Transport_ATS::Setup(const Teuchos::Ptr<State>& S)
 {
+
```
Comment (ecoon) [inline src/pks/transport/transport_ats_pk.cc:124]: extra spaces

File: src/pks/transport/transport_ats_pk.cc [hunk 24]
```diff
@@ -1184,6 +1203,7 @@ void Transport_ATS::AddMultiscalePorosity_(
 ******************************************************************* */
 void Transport_ATS::CommitStep(double t_old, double t_new, const Teuchos::RCP<State>& S)
 {
+
```
Comment (ecoon) [inline src/pks/transport/transport_ats_pk.cc:1206]: extra spaces

File: src/pks/transport/transport_ats_pk.cc [hunk 25]
```diff
@@ -1219,9 +1239,11 @@ void Transport_ATS::AdvanceDonorUpwind(double dt_cycle)
 
   // We advect only aqueous components.
   int num_advect = num_aqueous;
+
   int num_components = tcc_next.NumVectors();
   conserve_qty_->PutScalar(0.);
 
+
```
Comment (ecoon) [inline src/pks/transport/transport_ats_pk.cc:1246]: extra spaces

File: src/pks/transport/transport_ats_ti.cc [hunk 26]
```diff
@@ -124,6 +124,7 @@ void Transport_ATS::FunctionalTimeDerivative(double t,
     ComputeAddSourceTerms(t, 1., f_component, current_component_, current_component_);
   }
 
+
```
Comment (ecoon) [inline src/pks/transport/transport_ats_ti.cc:127]: extra spaces

**Review by ecoon — CHANGES_REQUESTED** [review]
Just a few minor things, the ColIterator, no "/" in units, and some more whitespace.

If you can revert files that have no real changes, that would be better.  There are several branches out that make changes in these files, and merging will be easier in the future for those branches if whitespace changes aren't introduced.

File: src/pks/flow/richards_pk.cc (hunk 16 above)
Comment (dasvyat) [reply src/pks/flow/richards_pk.cc]: fixed
